### PR TITLE
Refactor EnMasse installation

### PR DESCRIPTION
**Pull request by lulf**
* Remove use of jsonnet for templating
* Apply cluster role bindings and role bindings using kubernetes resources
* Read authentication service config dynamically
* Auto-generate config on controller startup (if needed)
* Update ansible playbooks to deploy using resources
* Generate bundles with predefined configuration examples

Issue #1623

### CODE REVIEW

**Comment by calohmn** [comment]
One question regarding K8s deployment: The old `deploy.sh` is still referenced from the top-level Makefile and from the Kubernetes install docs. For the next EnMasse version should there still be a `deploy.sh` for Kubernetes? Or for now only the manual install steps and then an ansible solution also for Kubernetes?

**Comment by lulf** [comment]
@calohmn 

The intention is to remove the deploy.sh, since it does not perform much useful function anymore other than to run openssl to create certs for api-server and auth service, before running `kubectl create -f examples/bundles/enmasse-with-standard-authservice`. Actually most of the functionality that was inside the ansible playbooks are now moved into the components themselves. This simplifies the ansible playbooks a great deal as well.

Since most users tend to write their own deploy scripts with their own customizations, I think the value of keeping it is very low right now. With the new templates, I think it will be easier to integrate with helm, operator framework etc.

File: documentation/common/configuring-using-cli.adoc [hunk 1]
```diff
@@ -2,9 +2,8 @@
 
 == Configuring {ProductName} using the command line
```
Comment (jenmalloy) [inline documentation/common/configuring-using-cli.adoc]: I think I moved this file in PR 1660 and made some edits to the wording

File: documentation/getting_started/configuring-address-spaces.adoc [hunk 2]
```diff
@@ -0,0 +1,38 @@
+[id='getting-started-configuring-address-spaces-{context}']
+= Configuring Address Spaces
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-address-spaces.adoc]: Configuring address spaces

File: documentation/getting_started/configuring-address-spaces.adoc [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+{ProductName} support the `brokered` and `standard` address spaces. This guide will use the
+`standard` address space. See <<address_model>> for more information about the differences.
+
+.Procedure 
+
+. Create an address space definition
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-address-spaces.adoc]: definition:

File: documentation/getting_started/configuring-address-spaces.adoc [hunk 4]
```diff
@@ -0,0 +1,38 @@
+[id='getting-started-configuring-address-spaces-{context}']
+= Configuring Address Spaces
+
+{ProductName} support the `brokered` and `standard` address spaces. This guide will use the
+`standard` address space. See <<address_model>> for more information about the differences.
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-address-spaces.adoc]: For more information about the differences, see xref:con-address-space-messaging[].

File: documentation/getting_started/configuring-address-spaces.adoc [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+. Create an address space definition
++
+[source,yaml,options="nowrap"]
+.link:standard-address-space.yaml[standard-address-space.yaml]
+----
+include::standard-address-space.yaml[]
+----
+
+. Create the address space
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-address-spaces.adoc]: space:

File: documentation/getting_started/configuring-address-spaces.adoc [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
++
+[options="nowrap",subs="attributes"]
+----
+{cmdcli} create -f standard-address-space.yaml
+----
++
+The address space is ready for use when the field `.status.isReady` is set to `true`.
+
+. Check status of address space
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-address-spaces.adoc]: Check the status of the address space:

File: documentation/getting_started/configuring-address-spaces.adoc [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
++
+[options="nowrap",subs="attributes"]
+----
+{cmdcli} get addressspace myspace -o jsonpath={.status.isReady}
+----
+
+. Retrieve console URL
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-address-spaces.adoc]: Retrieve the console URL:

File: documentation/getting_started/configuring-addresses.adoc [hunk 8]
```diff
@@ -1,69 +1,40 @@
-== Address types
+[id='getting-started-configuring-addresses-{context}']
+= Configuring Addresses
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-addresses.adoc]: Configuring addresses

File: documentation/getting_started/configuring-addresses.adoc [hunk 9]
```diff
@@ -1,69 +1,40 @@
...
-
-To deploy this configuration, you must currently use a http client like curl:
-
-[subs=attributes+,options="nowrap"]
-....
-curl -X POST -H "content-type: application/json" --data-binary @addresses.json -k https://$({OcGetRoute} restapi)/apis/enmasse.io/v1alpha1/namespaces/[:namespace]/addressspaces/default/addresses
-....
-
-This will connect to the address controller REST API and deploy the address config in the 'default' address space.
+See the <<address_model>> for details.  {ProductName} also comes with a console that you can use for managing addresses.
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-addresses.adoc]: For more details see xref:con-standard-address-space-messaging[].

File: documentation/getting_started/configuring-addresses.adoc [hunk 10]
```diff
@@ -1,69 +1,40 @@
...
+
+{ProductName} supports managing addresses natively in {PlatformName}.
+
+.Procedure
+
+. Create an address list definition
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-addresses.adoc]: definition:

File: documentation/getting_started/configuring-addresses.adoc [hunk 11]
```diff
@@ -1,69 +1,40 @@
...
++
+[source,yaml,options="nowrap"]
+.link:standard-addresses.yaml[standard-addresses.yaml]
+----
+include::standard-addresses.yaml[]
+----
+
+. Create the addressses
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-addresses.adoc]: addresses:

File: documentation/getting_started/configuring-addresses.adoc [hunk 12]
```diff
@@ -1,69 +1,40 @@
...
++
+[options="nowrap",subs="attributes"]
+----
+{cmdcli} create -f standard-addresses.yaml
+----
++
+The addresses are ready when the field `.status.isReady` is set to true.
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-addresses.adoc]: when the `.status.isReady` field is set to `true`.

File: documentation/getting_started/configuring-addresses.adoc [hunk 13]
```diff
@@ -1,69 +1,40 @@
...
+
+. Check status of addressses
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-addresses.adoc]: Check the status of the addresses:

File: documentation/getting_started/configuring-address-spaces.adoc [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+----
+
+. Create the address space
++
+[options="nowrap",subs="attributes"]
+----
+{cmdcli} create -f standard-address-space.yaml
+----
++
+The address space is ready for use when the field `.status.isReady` is set to `true`.
```
Comment (jenmalloy) [inline documentation/getting_started/configuring-address-spaces.adoc]: when the `.status.isReady` field is set to `true`.

File: documentation/getting_started/kubernetes.adoc [hunk 15]
```diff
@@ -1,52 +1,48 @@
...
 
-include::install-procedure.adoc[leveloffset+=1] 
+.Procedure 
 
-This guide uses a shell script for deploying {ProductName}. Windows users are advised to look at
-<<installing-kubernetes>>.
+. Download one of the releases from https://github.com/EnMasseProject/enmasse/releases and unpack it.
 
-include::deploy-procedure.adoc[leveloffset+=1] 
+. Create the namespace where you want to deploy {ProductName}
```
Comment (jenmalloy) [inline documentation/getting_started/kubernetes.adoc]: {ProductName}:

File: documentation/getting_started/kubernetes.adoc [hunk 16]
```diff
@@ -1,52 +1,48 @@
...
++
+[options="nowrap",subs="attributes"]
+----
+kubectl create namespace enmasse
+kubectl config set-context $(kubectl config current-context) --namespace=enmasse
+----
 
-[[role-based-access-control]]
-==== Role Based Access Control (RBAC)
+. Replace namespace references in bundle
```
Comment (jenmalloy) [inline documentation/getting_started/kubernetes.adoc]: Replace the namespace references in bundle:

File: documentation/getting_started/kubernetes.adoc [hunk 17]
```diff
@@ -1,52 +1,48 @@
...
+[options="nowrap",subs="attributes"]
+----
+sed -i "s/namespace: .*/namespace: enmasse/" examples/bundles/enmasse-with-none-authservice/*.yaml
+----
 
-The Kubernetes deployment script and YAML files currently do not support Role
-Based Access Control (RBAC). In Kubernetes clusters which have RBAC enabled, it is
-required to additionally create a role binding for the `default` service account
-to the `edit` role and for the `enmasse-admin` to the `cluster-admin` and `admin` roles:
+. Deploy using `enmasse-with-none-authservice` template
```
Comment (jenmalloy) [inline documentation/getting_started/kubernetes.adoc]: using the `enmasse-with-none-authservice` template:

File: documentation/getting_started/openshift.adoc [hunk 18]
```diff
@@ -1,33 +1,44 @@
...
 
-include::install-procedure.adoc[leveloffset+=1] 
+.Procedure 
 
-This guide uses a shell script for deploying {ProductName}. Windows users are advised to look at
-<<installing-openshift>>.
+. Download one of the releases from https://github.com/EnMasseProject/enmasse/releases and unpack it.
 
-include::deploy-procedure.adoc[leveloffset+=1] 
+. Login as as a user with cluster-admin
```
Comment (jenmalloy) [inline documentation/getting_started/openshift.adoc]: Log in as a user with `cluster-admin` privileges:

File: documentation/getting_started/openshift.adoc [hunk 19]
```diff
@@ -1,33 +1,44 @@
...
+. Login as as a user with cluster-admin 
++
+[options="nowrap",subs="attributes"]
+----
+oc login -u system:admin
+----
 
-[[configuring-addresses-openshift]]
-== Configuring addresses
+. Create the project where you want to deploy {ProductName}
```
Comment (jenmalloy) [inline documentation/getting_started/openshift.adoc]: {ProductName}:

File: documentation/getting_started/openshift.adoc [hunk 20]
```diff
@@ -1,33 +1,44 @@
...
++
+[options="nowrap",subs="attributes"]
+----
+oc new-project enmasse
+----
+
+. Deploy using `enmasse-with-none-authservice` template
```
Comment (jenmalloy) [inline documentation/getting_started/openshift.adoc]: using the `enmasse-with-none-authservice` template:

File: documentation/modules/proc-deploying-address-space-controller.adoc [hunk 21]
```diff
@@ -7,38 +7,20 @@
...
-[options="nowrap",subs="attributes"]
-----
-{cmdcli} create -f ./resources/resource-definitions/resource-definitions.yaml
-{cmdcli} create -f ./resources/plans/standard-plans.yaml
-{cmdcli} create -f ./resources/plans/brokered-plans.yaml
-----
+.Procedure
 
-. Deploy the address space controller:
+. Deploy the Address Space Controller
```
Comment (jenmalloy) [inline documentation/modules/proc-deploying-address-space-controller.adoc]: Controller:

File: documentation/modules/proc-deploying-api-server.adoc [hunk 22]
```diff
@@ -7,46 +7,31 @@
...
 * To install the {ProductName} API Server on {PlatformName}, you must have `cluster-admin` access to set up the
 required roles for delegating authentication to the {PlatformName} master; otherwise, you are restricted
 to using templates for creating address spaces. For more information about how to deploy without `cluster-admin` access,
 which restricts {ProductName} to using templates, see xref:installing-without-cluster-admin-privileges-{context}[].
 endif::[]
-
-.Procedure
-
-. Create a service account for the {ProductName} API server:
+* Modify the installation files according to the namespace the API Server is going to be installed in
```
Comment (jenmalloy) [inline documentation/modules/proc-deploying-api-server.adoc]: Modify the installation files to reflect the namespace in which the API Server will be installed:

File: documentation/modules/proc-deploying-api-server.adoc [hunk 23]
```diff
@@ -7,46 +7,31 @@
...
-+
-[options="nowrap"]
-----
-oc login -u system:admin
-oc create -f ./resources/cluster-roles/api-server.yaml
-----
+.Procedure
 
-. Grant privileges to the service account:
+. Register API service
```
Comment (jenmalloy) [inline documentation/modules/proc-deploying-api-server.adoc]: Register the API service:

File: documentation/modules/proc-deploying-api-server.adoc [hunk 24]
```diff
@@ -61,37 +46,11 @@ openssl req -new -x509 -batch -nodes -days 11000 -subj "/O=io.enmasse/CN=api-ser
...
 
-. Create the API server configuration:
-+
-[options="nowrap",subs="attributes"]
-----
-{cmdcli} create configmap api-server-config --from-literal=enableRbac=false
-----
-
-. Deploy the API server:
+. Deploy the API Server
```
Comment (jenmalloy) [inline documentation/modules/proc-deploying-api-server.adoc]: Server:

File: documentation/modules/proc-deploying-auth-services.adoc [hunk 25]
```diff
@@ -26,19 +27,29 @@ openssl req -new -x509 -batch -nodes -days 11000 -subj "/O=io.enmasse/CN=none-au
...
-{cmdcli} create -f ./resources/none-authservice/service.yaml
-{cmdcli} create -f ./resources/none-authservice/deployment.yaml
+{cmdcli} create -f ./examples/install/none-authservice
 ----
 
 == Deploying the `standard` authentication service
 
+.Prerequisites
+
+* Modify the installation files according to the namespace the Standard Authservice is going to be installed in
```
Comment (jenmalloy) [inline documentation/modules/proc-deploying-auth-services.adoc]: Modify the installation files to reflect the namespace in which the Standard Authservice will be installed:

File: documentation/modules/proc-deploying-service-broker.adoc [hunk 26]
```diff
@@ -1,119 +1,41 @@
 // Module included in the following assemblies:
 //
 // assembly-installing-manual-steps.adoc
-
+ifeval::["{cmdcli}" == "oc"]
 [id='deploying-service-broker-{context}']
 = (Optional) Deploying the service broker
 
-The service broker provides an implementation of the Open Service Broker API that integrates with the Kubernetes Service Catalog. The service broker requires the standard authentication service to be deployed.
+The service broker provides an implementation of the Open Service Broker API that integrates with the {PlatformName} Service Catalog. The service broker requires the standard authentication service to be deployed.
```
Comment (jenmalloy) [inline documentation/modules/proc-deploying-service-broker.adoc]: `standard`

File: documentation/modules/proc-deploying-service-broker.adoc [hunk 27]
```diff
@@ -1,119 +1,41 @@
...
 
-ifeval::["{PlatformName}" == "OpenShift"]
 .Prerequisites
-
 * The service broker requires the standard authentication service to be deployed.
```
Comment (jenmalloy) [inline documentation/modules/proc-deploying-service-broker.adoc]: `standard`

File: documentation/modules/proc-deploying-service-broker.adoc [hunk 28]
```diff
@@ -1,119 +1,41 @@
...
-oc policy add-role-to-user admin system:serviceaccount:enmasse:enmasse-admin
-oc adm policy add-cluster-role-to-user enmasse.io:service-broker system:serviceaccount:enmasse:enmasse-admin
-oc adm policy add-cluster-role-to-user system:auth-delegator system:serviceaccount:enmasse:enmasse-admin
-----
-+
-NOTE: You can log in again as the regular user after this step.
-endif::[]
-
-. Create a certificate to use for the service broker. For testing purposes, you can create a self-signed certificate:
+* Modify the installation files according to the namespace the Service Broker is going to be installed in
```
Comment (jenmalloy) [inline documentation/modules/proc-deploying-service-broker.adoc]: Modify the installation files to reflect the namespace in which the Service Broker will be installed:

File: documentation/modules/proc-deploying-service-broker.adoc [hunk 29]
```diff
@@ -1,119 +1,41 @@
...
-. Create a secret containing the service broker secret configuration:
-+
-[options="nowrap",subs="attributes"]
-----
-{cmdcli} create secret tls service-broker-secret --from-literal=keycloak.username=admin --from-literal=keycloak.password=admin --from-literal=keycloakCa.crt=`{cmdcli} extract secret/standard-authservice-cert --keys=tls.crt --to=-`
-----
+.Procedure
 
-. Deploy the service broker:
+. Deploy the Service Broker
```
Comment (jenmalloy) [inline documentation/modules/proc-deploying-service-broker.adoc]: Broker:

File: documentation/modules/proc-deploying-service-broker.adoc [hunk 30]
```diff
@@ -1,119 +1,41 @@
...
 +
 [options="nowrap",subs="attributes"]
 ----
-{cmdcli} create -f ./resources/service-broker/deployment.yaml
-{cmdcli} create -f ./resources/service-broker/service.yaml
+{cmdcli} apply -f examples/install/service-broker
 ----
 
-. To ensure the service broker redirects correctly, you must specify the `keycloakUrl` setting. If you are running a local cluster without a public DNS, use the internal service IP address for the host name; otherwise, use the host name of the external route. To obtain the service IP address, use this command:
+. Register the Service Broker with the OpenShift Service Catalog
```
Comment (jenmalloy) [inline documentation/modules/proc-deploying-service-broker.adoc]: Catalog:

**Comment by calohmn** [comment]
@lulf yes, makes sense.

Two more things here:
Concerning the replacement of namespaces, e.g.
````
sed -i "s/namespace: .*/namespace: enmasse/" examples/bundles/enmasse-with-none-authservice/*
````
mentioned in "A.2.1. Installing EnMasse".
In this case there is also "namespace: kube-system" in `examples/bundles/enmasse-with-none-authservice/030-RoleBinding-apiserver-authentication-reader.yaml` being replaced.
In general, I think replacing the string "myproject" could be a better idea, since it still appears in some places after the namespaces have been replaced.

Concerning the naming of the "examples" folder:
This I find a little bit confusing - I wouldn't look in such a folder for the deployment resources (although I understand that the files in there are meant to be customized for each installation, so in that sense they are indeed examples).
How about naming the folders something like this:
* install - or resources (was "examples")
  * bundles - or "as_bundles"
  * components  - or "by_components" (was "install")
  * templates  - or "as_templates", or "all-in-one"

**Comment by lulf** [comment]
> In general, I think replacing the string "myproject" could be a better idea, since it still appears in some places after the namespaces have been replaced

I agree, I will make that change to the docs

> 
> Concerning the naming of the "examples" folder:
> This I find a little bit confusing - I wouldn't look in such a folder for the deployment resources (although I understand that the files in there are meant to be customized for each installation, so in that sense they are indeed examples).
> How about naming the folders something like this:
> 
> * install - or resources (was "examples")
>   
>   * bundles - or "as_bundles"
>   * components  - or "by_components" (was "install")
>   * templates  - or "as_templates", or "all-in-one"

I agree, I think the renaming `examples` to `install` and `examples/install` to `install/components` makes sense. Wdyt @grs @k-wall @vbusch @robshelly ?

The paths will then be:

`install/bundles`
`install/components`
`install/templates`

**Pull request merged**